Set Outpost value for TownPreClaimEvent

Travja opened the pull request:
#### Description: 
Essentially, the TownPreClaimEvent's TownBlock#isOutpost method would always return false. This pull request fixes the issue, so plugins using the event can properly address if the block is an outpost.
___
- [x] I have tested this pull request for defects on a server. 

By making this pull request, I represent that I have the right to waive copyright and related rights to my contribution, and agree that all copyright and related rights in my contributions are waived, and I acknowledge that the TownyAdvanced organization has the copyright to use and modify my contribution under the Towny [License](https://github.com/LlmDl/Towny/blob/master/LICENSE.md) for perpetuity.

silverwolfg11 commented:
It would be nice if the change events were cancellable. Also there are some places where those changes are implicit that you may want to account for. E.g. when the nation capital changes its mayor, the nation's leader is changed. When the nation changes its capital, it's leader is changed.

Travja commented:
Awesome. I can do that. Should I throw an error if it's canceled? If so, how do you want to handle the language options for the error message?

suneettipirneni commented:
> Awesome. I can do that. Should I throw an error if it's canceled? If so, how do you want to handle the language options for the error message?

@Travja Don't worry about the lang strings, we can do that part, and usually, when an event is cancelled in towny the event has a `cancelMessage` field the is defined in the constructor. So other plugins can change the message as they please.

Travja commented:
I've added the various other calls to the events, and added a NationCapitalChangeEvent to catch when the nations change capitals. Hopefully I've covered all the bases for when those should be called. I also added some lang entries for English, as well as adding a forceSetCapital method in Nation that doesn't throw an exception and doesn't allow for the NationCapitalChangeEvent or NationKingChangeEvent to be cancelled.

LlmDl commented on src/com/palmergames/bukkit/towny/object/Nation.java:
```diff
@@ -4,9 +4,7 @@
-import com.palmergames.bukkit.towny.event.NationAddTownEvent;
-import com.palmergames.bukkit.towny.event.NationRemoveTownEvent;
-import com.palmergames.bukkit.towny.event.NationTagChangeEvent;
+import com.palmergames.bukkit.towny.event.*;
```
No wildcard imports please.

LlmDl commented on src/com/palmergames/bukkit/towny/object/Town.java:
```diff
@@ -4,9 +4,7 @@
-import com.palmergames.bukkit.towny.event.TownAddResidentEvent;
-import com.palmergames.bukkit.towny.event.TownRemoveResidentEvent;
-import com.palmergames.bukkit.towny.event.TownTagChangeEvent;
+import com.palmergames.bukkit.towny.event.*;
```
No wildcard imports please.

silverwolfg11 commented on src/com/palmergames/bukkit/towny/object/Nation.java:
```diff
@@ -204,7 +206,32 @@ else if (town.hasNation())
+			return;
+		
+		Bukkit.getPluginManager().callEvent(new NationCapitalChangeEvent(this, capital, false));
+		Bukkit.getPluginManager().callEvent(new NationKingChangeEvent(this, capital.getMayor(), false));
```
Since these methods have pretty much the same code, I would overload them so that one method accepts a boolean cancellable parameter, while the other is a default method that calls the overloaded method with cancellable passed in as true.
E.g.
```java
setCapital(Town capital) {
    setCapital(capital, true);
}

setCapital(Town capital, boolean cancellable) {
...
}
```

Travja replied:
I've implemented it, but with the default being a non-forced call. Let me know your thoughts on that.

silverwolfg11 commented on src/com/palmergames/bukkit/towny/event/NationKingChangeEvent.java:
```diff
@@ -0,0 +1,66 @@
+
+	@Override
+	public boolean isCancelled() {
+		return isCancellable ? isCancelled : isCancellable;
```
Can be changed to `return isCancellable && isCancelled;`

silverwolfg11 commented:
Sorry for such a late response. 
One thing that has been making me hesitant on this PR was the `NationCapitalChangeEvent` implementing `cancellable` but having the potential to not be able to be cancelled. This would not be a good design decision because when an event implements the cancellable interface, there is a sort of contract between the Bukkit API and plugins that when a plugin cancels the event, the action that the event is displaying will not progress (or in some cases be reverted).
That led me to look at the usages of when set capital is being force called. These would be in areas of load, deletes, and proximity checks.  After some time, I realized that while it would be less code to implement the change event at the base level `setCapital()` method in the nation class (like this PR has done) it offers us less control to change the outcome.
Thus what I think may be a better solution is moving the event calls (at least the `NationChangeKingEvent`) to higher calls such as within the command classes themselves where we can control a cancelled event outcome better. 
Again, sorry for such a late response, and I appreciate all the hard-work that you put into this PR! 😄

Travja commented:
No worries! I totally understand that concern. It was a bit weird for me putting it in in the first place. What if instead of having the NationCapitalChangeEvent implement cancellable, we had a PreNationCapitalChangeEvent, following the pattern for other events, that did implement cancellable. This event would be called in the higher call sections and dependent on the outcome of NationChangeKingEvent, but wouldn't be called when the plugin is specifically doing it's background logic as those aren't areas where it would make sense to  allow cancelling the event in the first place.

silverwolfg11 commented:
Sounds great, tysm!
